server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1433,7 +1433,12 @@ private void innerOpenEngineAndTranslog() throws IOException {
         final String translogUUID = store.readLastCommittedSegmentsInfo().getUserData().get(Translog.TRANSLOG_UUID_KEY);
         final long globalCheckpoint = Translog.readGlobalCheckpoint(translogConfig.getTranslogPath(), translogUUID);
         replicationTracker.updateGlobalCheckpointOnReplica(globalCheckpoint, "read from translog checkpoint");
-        updateRetentionLeasesOnReplica(loadRetentionLeases());
+        // only recover retention leases if recovering from an existing store _and_ not bootstrapping a new history UUID.
+        if (recoveryState.getRecoverySource() == RecoverySource.ExistingStoreRecoverySource.INSTANCE) {
```

Comment:
This logic should preferably go into `StoreRecovery`, where we also do the initialization of the lucene commit and translog files. Whenever we call `createEmptyTranslog` we should also delete the retention leases file.

Reply:
+1

Reply:
Ok. By correlating this with `createEmptyTranslog` do you mean that we should consider a missing retention leases state file as an error when loading it? Today we treat this as if it were empty.

Reply:
Yeah, I think we should not be lenient if we prepare the lease file before opening an Engine.

Reply:
Ok, we can't be strict here yet because of BWC: indices created in 7.x do not yet have a retention lease file, and indices created in 6.x (<6.7.0) will never do so. I left the lenience in and will remove it in a followup after backporting. I included an assertion to ensure we follow up.